File: app/src/main/java/yagoc/pieces/Rook.java
package yagoc.pieces;

import yagoc.board.BoardView;
import yagoc.board.Move;
import yagoc.board.Square;

import java.util.stream.Stream;

public class Rook extends Piece {
    public Rook(PieceColor pieceColor) {
        super(PieceType.Rook, pieceColor);
    }

    static boolean isCorrectMoveForRook(BoardView board, Move move) {
        if (move.hasSameRank()) {
            //movimiento horizontal
            int mi = Math.min(move.from().file(), move.to().file()) + 1;
            int ma = Math.max(move.from().file(), move.to().file());
            for (; mi < ma; mi++) {
                if (!board.pieceAt(move.from().rank(), mi).equals(Pieces.none))
                    return false;
            }
            return true;
        } else if (move.hasSameFile()) {
            //movimiento vertical
            int mi = Math.min(move.from().rank(), move.to().rank()) + 1;
            int ma = Math.max(move.from().rank(), move.to().rank());
            for (; mi < ma; mi++) {
                if (!board.pieceAt(mi, move.from().file()).equals(Pieces.none))
                    return false;
            }
            return true;
        } else
            return false;
    }

    static Stream<Move> generateMovesForRook(BoardView board, Square from) {
        Piece piece = board.pieceAt(from);
        return from.straightSquares().stream()
                .map((to) -> new Move(piece, from, to));
    }

    @Override
    public boolean isValidForPiece(BoardView board, Move move) {
        return isCorrectMoveForRook(board, move);
    }

    @Override
    public Stream<Move> generateMovesForPiece(BoardView board, Square from) {
        return generateMovesForRook(board, from);
    }
}

File: app/src/main/kotlin/yagoc/pieces/Rook.kt
package yagoc.pieces

import yagoc.board.BoardView
import yagoc.board.Move
import yagoc.board.Square
import yagoc.board.square
import kotlin.math.max
import kotlin.math.min

class Rook(pieceColor: PieceColor) : Piece(PieceType.Rook, pieceColor) {
    public override fun isValidForPiece(board: BoardView, move: Move): Boolean = isCorrectMoveForRook(board, move)

    public override fun generateMovesForPiece(board: BoardView, from: Square) = generateMovesForRook(board, from)

    companion object {
        @JvmStatic
        fun isCorrectMoveForRook(board: BoardView, move: Move): Boolean {
            return when {
                move.hasSameRank() -> {
                    // move horizontally
                    var mi = min(move.from.file, move.to.file) + 1
                    val ma = max(move.from.file, move.to.file)
                    while (mi < ma) {
                        if (board.someAt(square(move.from.rank, mi))) return false
                        mi++
                    }
                    true
                }
                move.hasSameFile() -> {
                    // move vertically
                    var mi = min(move.from.rank, move.to.rank) + 1
                    val ma = max(move.from.rank, move.to.rank)
                    while (mi < ma) {
                        if (board.someAt(square(mi, move.from.file))) return false
                        mi++
                    }
                    true
                }
                else -> false
            }
        }

        @JvmStatic
        fun generateMovesForRook(board: BoardView, from: Square) =
            board.pieceAt(from).let { piece ->
                from.straightSquares().stream()
                    .map { to: Square -> Move(piece, from, to) }
            }
    }
}